---
title: "hw_3"
author: "part II"
date: "22/02/21"
output: 
  html_document:
    toc: true 
    toc_depth: 4  
    theme: united  
    highlight: tango  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

  
# Before data analise

For package installation we used function from - <https://stackoverflow.com/a/31620488>. 

It will check for missing packages and install them + it will query both - CRAN and Bioconductor repositories for missing package.


```{r}
install.packages.auto <- function(x) { 
  x <- as.character(substitute(x)) 
  if(isTRUE(x %in% .packages(all.available=TRUE))) { 
    eval(parse(text = sprintf("require(\"%s\")", x)))
  } else { 
    #update.packages(ask= FALSE) #update installed packages.
    eval(parse(text = sprintf("install.packages(\"%s\", dependencies = TRUE)", x)))
  }
  if(isTRUE(x %in% .packages(all.available=TRUE))) { 
    eval(parse(text = sprintf("require(\"%s\")", x)))
  } else {
    source("http://bioconductor.org/biocLite.R")
    #biocLite(character(), ask=FALSE) #update installed packages.
    eval(parse(text = sprintf("biocLite(\"%s\")", x)))
    eval(parse(text = sprintf("require(\"%s\")", x)))
  }
}
```

## Install and library all packages.

```{r message=FALSE}
install.packages.auto("dplyr")
install.packages.auto("readxl")
install.packages.auto("ggplot2")
install.packages.auto("hrbrthemes")
# BiocManager::install("Hmisc")
library("Hmisc")
install.packages.auto("car")
install.packages.auto("gplots")
install.packages.auto("multcomp")
```

# Used notation: 

Genotype:

* Control = normal mice

* Ts65Dn - model mice. In Ts65Dn, cells possess an extra copy of a segment of genes on chromosome 16 as well as a segment of genes on chromosome 17. From this model, various Down syndrome phenotypes are produced, including behavioral abnormalities and cognitive defects.


Bechavior:

* Stimulated to learn mice - C-S

* Not stimulated to learn mice - S-C

Treatment

* memantine

* saline (= no memantine)


# EDA


```{r}
DCN <- read_excel("Data_Cortex_Nuclear.xls")
```


```{r}
str(DCN)
```


## 1.1 Measurements and mouses

```{r}
# 1.1


# measurements in study
n_measurements <- length(unique(DCN$MouseID))

# mouses in study
mouse_mesur_list <- strsplit(DCN$MouseID, "_")
df <- data.frame(matrix(unlist(mouse_mesur_list), nrow=length(mouse_mesur_list), byrow=TRUE))

df$X2 <- as.numeric(df$X2)

n_mouse <- length(unique(df$X1))

# measurements per mouse ?
table(df$X1)
unique(table(df$X1))

```

### result
There are in total `r n_measurements` measurements in study on `r n_mouse` mouses. 

One mouse -- `r unique(table(df$X1))` measurements. 


## 1.2 and 1.3 Groups

```{r}
# create dataset with groups only
DCN_groups <- data.frame(mouse_id_num = df$X1, 
                         genotype = as.factor(DCN$Genotype), 
                         treatment = as.factor(DCN$Treatment), 
                         behavior = as.factor(DCN$Behavior), 
                         class = as.factor(DCN$class))

# uniq vlues
# unique(DCN_groups$genotype)
# unique(DCN_groups$treatment)
# unique(DCN_groups$behavior)
# unique(DCN_groups$class)

# there are duplicates
# sum(duplicated(DCN_groups))

# delete duplicates
DCN_groups_uniq <- DCN_groups %>% distinct()

# groups
## genotype
g <- table(DCN_groups_uniq$genotype) 
g

## treatment
t <- table(DCN_groups_uniq$treatment) 
t

## behavior
b <- table(DCN_groups_uniq$behavior) 
b

## genotype + treatment
g_t <- table(DCN_groups_uniq$genotype, DCN_groups_uniq$treatment) 
g_t

## genotype + treatment + behavior
g_t_b <- table(DCN_groups_uniq$genotype, DCN_groups_uniq$treatment, DCN_groups_uniq$behavior) 
g_t_b


## class = genotype + treatment + behavior
c <- table(DCN_groups_uniq$class) 
c



cl <- length(unique(DCN_groups_uniq$class))
```


### result

Split by one feature:

* There were mice with normal genotype (control) and with trisomy - `r g`. 

* There were mice treated with Memantine and treated with saline - `r t`. 

* There were mice stimulated to learn and NOT stimulated to learn - `r b`. 


Split by two features:

* From control mice `r g_t[1]` were treated with Memantine and `r g_t[3]` were treated with saline.

* From trisomy mice `r g_t[2]` were treated with Memantine and `r g_t[4]` were treated with saline.


Split by tree features:

* From control mice which were treated with Memantine, `r g_t_b[1]` were stimulated to learn.

* From control mice which were treated with Memantine, `r g_t_b[5]` were NOT stimulated to learn.

* From trisomy mice which were treated with Memantine, `r g_t_b[2]` were stimulated to learn.

* From trisomy mice which were treated with Memantine, `r g_t_b[6]` were NOT stimulated to learn.


* From control mice which were treated with Saline, `r g_t_b[3]` were stimulated to learn.

* From control mice which were treated with Saline, `r g_t_b[7]` were NOT stimulated to learn.

* From trisomy mice which were treated with Saline, `r g_t_b[4]` were stimulated to learn.

* From trisomy mice which were treated with Saline, `r g_t_b[8]` were NOT stimulated to learn.


**So there are `r cl` classes in experiment. These classes are not well balanced. **

* Max group size: `r max(c)` 
* Min group size: `r min(c)`  



## 1.4 Complete observations

### which rows contain NA?

```{r}
nrow(DCN) # total rows
nrow(DCN[!complete.cases(DCN),]) # rows with NA 528
nrow(DCN[complete.cases(DCN),]) # rows without NA 552


```

### which cols contain NA?

```{r}

# which cols contain NA?
cols_NA <- sapply(DCN, function(x) sum(is.na(x)))

# how many cols contain NA?

length(cols_NA[cols_NA != 0])


# names of cols
names_cols_NA <- names(cols_NA)

# unique quantities of NA 
unique(cols_NA)

# 0 NA
zero_na <- names_cols_NA[cols_NA == 0]
zero_na
# 3 NA
names_cols_NA[cols_NA == 3]
# 7 NA
names_cols_NA[cols_NA == 7]
# 18 NA
names_cols_NA[cols_NA == 18]
# 75 NA
names_cols_NA[cols_NA == 75]
# 180 NA
names_cols_NA[cols_NA == 180]
# 210 NA
names_cols_NA[cols_NA == 210]
# 213 NA
names_cols_NA[cols_NA == 213]
# 270 NA
names_cols_NA[cols_NA == 270]
# 285 NA
names_cols_NA[cols_NA == 285]

```

#### Result

**In dataset with `r nrow(DCN)` rows `r nrow(DCN[!complete.cases(DCN),])` include NA**

**Complete observations there are for proteins: `r zero_na[c(2:29)]`**


## Should we delete all NA rows?

```{r}
DCN_raw_no_NA <- na.omit(DCN)

mouse_mesur_list_raw_no_NA <- strsplit(DCN_raw_no_NA$MouseID, "_")

df_raw_no_NA <- data.frame(matrix(unlist(mouse_mesur_list_raw_no_NA), nrow=length(mouse_mesur_list_raw_no_NA), byrow=TRUE))

DCN_raw_no_NA_groups <- data.frame(mouse_id_num = df_raw_no_NA$X1, 
                         genotype = as.factor(DCN_raw_no_NA$Genotype), 
                         treatment = as.factor(DCN_raw_no_NA$Treatment), 
                         behavior = as.factor(DCN_raw_no_NA$Behavior), 
                         class = as.factor(DCN_raw_no_NA$class))

DCN_raw_no_NA_groups <- DCN_raw_no_NA_groups %>% distinct()
# groups
## genotype
g <- table(DCN_raw_no_NA_groups$genotype) 
g

## treatment
t <- table(DCN_raw_no_NA_groups$treatment) 
t

## behavior
b <- table(DCN_raw_no_NA_groups$behavior) 
b

## genotype + treatment
g_t <- table(DCN_raw_no_NA_groups$genotype, DCN_raw_no_NA_groups$treatment) 
g_t

## class
c_raw_no_NA <- table(DCN_raw_no_NA_groups$class) 
c_raw_no_NA


### compare balance of classes in dataset with NA and without NA
df3 <- data.frame(c, c_raw_no_NA)
df3

max(c_raw_no_NA)
min(c_raw_no_NA)


```

**If we remove observations with missing values, the class sizes become very small and their imbalance increases.**

* Max group size: `r max(c_raw_no_NA)` 
* Min group size: `r min(c_raw_no_NA)`  


**The another way is to delete some cols**


## Should we delete some cols with NA ?


```{r}


# cols with many NA
cols_no_NA <- sort(cols_NA[cols_NA != 0 & cols_NA >= 100 ])
cols_no_NA
# gen col number of cols with many NA
which(names(DCN)%in%names(cols_no_NA))



```

**I will delete cols with NA >= 100.**

* BCL2_N 285 NA 

* H3MeK4_N 270 NA 

* BAD_N 213 NA

* EGR1_N 210 NA 

* H3AcK18_N 180 NA


```{r}
DCN_cols_no_NA <- DCN[, -c(70, 71, 75, 76, 77)]

nrow(DCN_cols_no_NA[!complete.cases(DCN_cols_no_NA),]) # rows with NA 528 in raw data
nrow(DCN_cols_no_NA[complete.cases(DCN_cols_no_NA),]) # rows without NA 552 in raw data

```
**In result i will get `r nrow(DCN_cols_no_NA[complete.cases(DCN_cols_no_NA),])` observations without NA**

So i remove rest observations with NA to get 972 observations. 


```{r}
DCN_cols_no_NA <- DCN_cols_no_NA[complete.cases(DCN_cols_no_NA),]
DCN_cols_no_NA
```



```{r}
mouse_mesur_list <- strsplit(DCN_cols_no_NA$MouseID, "_")

df <- data.frame(matrix(unlist(mouse_mesur_list), nrow=length(mouse_mesur_list), byrow=TRUE))

DCN_cols_no_NA_groups <- data.frame(mouse_id_num = df$X1, 
                         genotype = as.factor(DCN_cols_no_NA$Genotype), 
                         treatment = as.factor(DCN_cols_no_NA$Treatment), 
                         behavior = as.factor(DCN_cols_no_NA$Behavior), 
                         class = as.factor(DCN_cols_no_NA$class))

DCN_cols_no_NA_groups <- DCN_cols_no_NA_groups %>% distinct()


# for dataset, with 5 cols and 108 raws deleted
table(DCN_cols_no_NA_groups$class)

# for data with 528 raws deleted
table(DCN_raw_no_NA_groups$class) 

# for raw data
table(DCN_groups_uniq$class)

```

**As a result, in order to preserve the data structure, it seemed to me a better option to delete 5 columns (= proteins) for which there are no data for 20% of cases, and then delete the remaining observations (108), which have missing data.**


# BDNF_N

## 2

### pointrange plot

```{r}
DF <- DCN_cols_no_NA

DF$class <- as.factor(DF$class)

p <- ggplot(DF, aes(x = class, y = BDNF_N, color = class)) 

p + geom_boxplot() + theme_minimal()


```
```{r}
# median value in each class
a <- aggregate(DF$BDNF_N, by = list(DF$class), FUN=median)
a
```

**The standard deviations in all five groups were approximately the same, ranging from `r min(a$x)` to `r max(a$x)`**

### dispersal analize
```{r}
fit <- aov(data = DF, BDNF_N ~ class)
summary(fit)
```


**The F-test for different classes of mice (class) was statistically significant (p <.0001), which indicates a difference in the expression of the BDNF_N protein in mice from different classes.**



# Means and 95% Confidence Intervals for Eight Classes of Mice
```{r}
plotmeans(data = DF, BDNF_N ~ class, xlab='Mice-class', ylab='BDNF_N',
          main='Means and 95% Confidence Intervals for Eight Classes of Mice')
```
```{r}
tuk <- glht(fit, linfct=mcp(class='Tukey'))
summary(tuk)
```
```{r}
par(mar=c(5,5,8,2))

plot(cld(tuk, level=.05),
sub="Tukey test")
```

**Classes designated by the same letters do not differ statistically.**

# ERBB4_N

## 3.1



## 3.2


# PCA

## 4.1

## 4.2

## 4.3

## 4.4








































































